### DISCUSSION TITLE
Please provide some way to track asset usage

### DISCUSSION THREAD

**Discussion by darylknight** [craftcms/cms#11934] [Ideas] [answered]
One of the first questions most of our clients ask after having their site live for a little while is "How can I tell which assets are being used?" - so they know what can be deleted, or which of 8 identical images are actually being used after a designer uploaded multiple very similar versions of an illustration. This gets a lot worse the older a site gets, and have no way to track which images are still being used.

There's been a plugin around (Asset Usage) for a long time, but since about 3.2 (the big overhaul to revisions, live preview and drafts) it's never been able to accurately tell you whether an asset is used or not.

Would you consider adding something like this to Craft core? Specifically using Asset fields and relations - I know it'd be a pain in the ass to track things like images inside redactor fields, but anything would be better than nothing, given that we always disable the image button in Redactor.

**Comment by brandonkelly** [comment]
My concern is, without tracking usages within Redactor/CKEditor fields, users would start deleting assets that don’t _seem_ to be in use, which will end up breaking the Redactor/CKEditor fields.

I think the solution is that Redactor/CKEditor need to start storing relations for any elements they’re referencing, just like relational fields. Then we’d be able to accurately show all usages for any given asset.

**Reply by darylknight** [reply]
That sounds like a good idea. Is it likely to be implemented?

**Reply by brandonkelly** [reply]
Added to our 5.0 project! https://craftcms.com/roadmap

**Reply by gbowne-quickbase** [reply]
The https://plugins.craftcms.com/element-relations?craft4 by [internetztube Frederic Köberl](https://github.com/internetztube) already does look into redactor fields if you need it now

**Reply by mathijsbrand** [reply]
So what is the status for Craft CMS 5 on this one? 
I don't see this issue on the roadmap. And the elements-relations plugin doesn't work on Craft CMS 5.

**Reply by LucasPoignonnec** [reply]
Hello!

Are there any plans in the near future to make tracking asset usage part of Craft core?

We are in a similar case as the original author, currently using the plugin '[Asset Usage](https://plugins.craftcms.com/assetusage?craft4)', however it is no longer maintained according to the plugin store.
We also consider using the '[Element relations](https://plugins.craftcms.com/element-relations?craft4)' plugin, however this could also lose support in the future.

As this feature is requested by quite a lot of our clients, a more permanent solution would be amazing.

Thank you for your feedback.

**Comment by brandonkelly** [comment]
Getting closer to being able to resolve this in Craft 5.3, as we’re adding the framework for giving CKEditor fields the ability to store relations for any images they embed (#15400).

**Reply by kbergha** [reply]
It should add relations to any asset that is linked to (typically documents), not only those that are embedded.

**Reply by ryanleichty** [reply]
Just stumbled on this thread as we're still using the Asset Usage plugin. Is this something that's still planned to be added natively?

**Reply by MWhiteFearless** [reply]
Are there any upcoming plans to add features similar to the asset usage plugin now that there's support for it? I would really love this addition to Craft! The asset usage plugin unfortunately still doesn't work for my project (a known issue with postgresql)

Edit: the Asset Usage plugin released a fix for the postgres issue and it works now! I still think it would be nice to have as a core feature, but great to have the plugin working in the meantime

**Comment by yannkost** [comment]
[https://plugins.craftcms.com/asset-cleaner?craft5](https://plugins.craftcms.com/asset-cleaner?craft5) This may could be a solution? For When I made this I managed to get usage in CKEditor and Redactor fields shown as well.

**Comment by larsboldt** [comment]
@brandonkelly Any update on this feature? Would be nice if images and documents in wysiwygs are tracked with the relation table

**Comment by yannkost** [comment]
@larsboldt tracking this with the relation table is not easily feasable, each plugin for wysiwygs editors could/does handle it's content in it's own way and is much more complex that an asset/entry field. (It's not a "native" field type, also the content column in the element_sites table contains almost always all "html/markdown" content of such editor fields.)

**Reply by larsboldt** [reply]
P&T provides both ckeditor and the link plugin, thus have full control over its implementation. On save event, parse the wysiwyg data, look for their own macro, keep track in relation table? P&T could provide some interface or (macro)pattern for ckeditor plugins that allow other providers links to be catched by the same code?

**Comment by yingban** [comment]
@larsboldt 

In my project and asset config, I added a TWIG template which gives 1:1 relations between assets and the entry where it is used.
Since you can use inline entries in CKEditor, it is working for assets inside CKEditor.

sorry the code view was not satisfying, so I just put a screenshot
<img width="852" height="782" alt="Screenshot 2026-04-17 at 10 41 24" src="https://github.com/user-attachments/assets/8a3ea78b-4c6c-4ac7-9a13-ee032d2453a5" />

**Comment by brandonkelly** [comment, marked as answer]
Craft 5.10 is out with an improved element deletion workflow, which now shows a modal when elements are being deleted that are related by other entries. (#18728)
